### PR TITLE
[Profiler] Implement ICorProfilerCallback::EventPipeEventDelivered to receive CLR events synchronously

### PR DESCRIPTION
**Pull request by chrisnas**
Create a session to receive CLR events via ICorProfilerCallback
Handle AllocationTickV4 (Allocation profiling) and ContentionStopV1 (Contention profiling)

### CODE REVIEW

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp [hunk 1]
```diff
@@ -576,8 +618,36 @@ HRESULT STDMETHODCALLTYPE CorProfilerCallback::Initialize(IUnknown* corProfilerI
...
     // Log some more important environment info:
-    CorProfilerCallback::InspectRuntimeVersion(_pCorProfilerInfo);
+    USHORT major = 0;
+    USHORT minor = 0;
+    CorProfilerCallback::InspectRuntimeVersion(_pCorProfilerInfo, major, minor);
+
+    // CLR events-based profilers need ICorProfilerInfo12 (i.e. .NET 5+) to setup the communication.
+    // If no such provider is enabled, no need to trigger it.
+    // TODO: update the test when a new CLR events-based profiler is added (contention, GC, ...)
+    if ((major >= 5) && _pConfiguration->IsAllocationProfilingEnabled())
```
Comment (kevingosse) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp:631]: I think we should log a warning if `IsAllocationProfilingEnabled` is true but the runtime is not .NET 5+

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+    {
+        // Dump to see all received events
+        //DumpEvent(providerName, name, id, version, keywords, eventId, eventVersion, numStackFrames);
+    }
+}
+
+
+// dump the buffer every 16 bytes + corresponding ASCII characters
+// ex:  44 4F 54 4E 45 54 5F 49  50 43 5F 56 31 00 77 00    DOTNET_IPC_V1.w.
+const DWORD LineWidth = 16;
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:154]: avoid defining global variables in the middle of the file. prefer static member (either class, either function)
scoping them is better.

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+#include "shared/src/native-src/string.h"
+#include "assert.h"
+
+
+#define LONG_LENGTH 1024
+
+// from https://github.com/dotnet/samples/blob/main/core/profiling/eventpipe/src/CorProfiler.h
+template <class Key, class Value>
+class ThreadSafeMap
```
Comment (gleocadie) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h:26]: maybe put this in its own file

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/AllocationsProvider.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+#define CALL(x)                                                                                               \
+    {                                                                                                         \
+        HRESULT hr = x;                                                                                       \
+        if (FAILED(hr))                                                                                       \
+        {                                                                                                     \
+            Log::Warn("Profiler call failed with result ", HResultConverter::ToStringWithCode(hr), ": ", #x); \
+            return;                                                                                           \
+        }                                                                                                     \
+    }
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/AllocationsProvider.cpp:24]: Good idea.
Let's do that in another PR

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/AllocationsProvider.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                                       const WCHAR* typeName,
+                                       uintptr_t address,
+                                       uint64_t objectSize)
+{
+    // create a sample from the allocation
+
+    ManagedThreadInfo* threadInfo;
+    CALL(_pManagedThreadList->TryGetCurrentThreadInfo(&threadInfo))
+
+    const auto pStackFramesCollector = OsSpecificApi::CreateNewStackFramesCollectorInstance(_pCorProfilerInfo);
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/AllocationsProvider.cpp:52]: Good point for another PR

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp [hunk 6]
```diff
@@ -0,0 +1,347 @@
...
+ClrEventsParser::ClrEventsParser(ICorProfilerInfo12* pCorProfilerInfo, IAllocationsListener* pAllocationListener)
+    :
+    _pCorProfilerInfo{pCorProfilerInfo},
+    _pAllocationListener{pAllocationListener}
+{
+}
+
+const char* GetEventName(DWORD eventId)
+{
+    switch (eventId)
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:23]: I'm removing the traces that were used for investigation

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp (hunk 1 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp:631]: ✅ done

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp [hunk 7]
```diff
@@ -0,0 +1,347 @@
...
+    std::cout << std::hex;
+    for (DWORD i = 0; i < byteCount; i++)
+    {
+        std::cout << std::uppercase << std::setfill('0') << std::setw(2) << (int)pBuffer[i] << " ";
+        stringBuffer[pos] = GetCharFromBinary(pBuffer[i]);
+        pos++;
+
+        if (pos % LineWidth == 0)
+        {
+            std::cout << "    ";
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:179]: I'm removing the traces that were used in the console for investigation

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp [hunk 8]
```diff
@@ -0,0 +1,347 @@
...
+        //std::stringstream buffer;
+        //buffer << "ContentionStop";
+        //buffer << ((payload.ContentionFlags == 0) ? " - managed " : " - native ");
+        //buffer << "(" << payload.DurationNs << " ns)";
+        //buffer << std::endl;
+        //std::cout << buffer.str();
+    }
+}
+
+const std::string UnknownProvider("UnknownProvider");
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:307]: I'm removing the traces that were used in the console for investigation

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp [hunk 9]
```diff
@@ -0,0 +1,347 @@
...
+    {
+        WCHAR nameBuffer[LONG_LENGTH];
+        ULONG nameCount;
+        HRESULT hr = _pCorProfilerInfo->EventPipeGetProviderInfo(provider, LONG_LENGTH, &nameCount, nameBuffer);
+        if (FAILED(hr))
+        {
+            return UnknownProvider;
+        }
+
+        _providerNameCache.insertNew(provider, shared::ToString(shared::WSTRING(nameBuffer)));
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:322]: Thanks: I had a comment in the POC PR to remind me to ask you how to optimize this kind of things  ;^)
https://github.com/DataDog/dd-trace-dotnet/pull/2925/files#diff-9a090c5549e071d7097d2a93b3ae016f10381339b7aa7510ce296a9d5fc7cf56R321

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        return _map.find(key);
+    }
+
+    typename std::map<Key, Value>::const_iterator end() const
+    {
+        return _map.end();
+    }
+
+    // Returns true if new value was inserted
+    bool insertNew(Key key, Value value)
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h:45]: Same

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+            return false;
+        }
+
+        _map[key] = value;
+        return true;
+    }
+};
+
+struct AllocationTickV4Payload
+{
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h:60]: No required: fields are public by default in struct

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+                                   // this is the address of the MethodTable that corresponds to the last object allocated (the object that caused the 100 KB threshold to be exceeded).
+    const WCHAR* TypeName;         // The name of the type that was allocated. When there are several types of objects that were allocated during this event,
+                                   // this is the type of the last object allocated (the object that caused the 100 KB threshold to be exceeded).
+    uint32_t HeapIndex;            // The heap where the object was allocated. This value is 0 (zero) when running with workstation garbage collection.
+    uintptr_t Address;             // The address of the last allocated object.
+    uint64_t ObjectSize;           // The size of the last allocated object.
+};
+
+struct ContentionStopV1Payload
+{
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h:79]: same  :^)

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp [hunk 13]
```diff
@@ -124,6 +124,25 @@ bool CorProfilerCallback::InitializeServices()
...
+                _pCorProfilerInfo,
+                _pManagedThreadList,
+                _pFrameStore.get(),
+                _pThreadsCpuManager,
+                _pAppDomainStore.get(),
+                pRuntimeIdStore);
+        }
+
+        // TODO: add new CLR events-based providers to the event parser
+        _pClrEventsParser = new ClrEventsParser(_pCorProfilerInfoEvents, _pAllocationsProvider);
```
[the remark this answers is not in the document]
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/CorProfilerCallback.cpp:143]: ✅  Good catch!

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp (hunk 2 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.cpp:154]: I'm removing the traces that were used in the console for investigation

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h (hunk 3 above)
Comment (chrisnas) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/ClrEventsParser.h:26]: taken from MS sample : removing it with the rest of the traces

**Review by gleocadie — APPROVED** [review]
LGTM

**Pull request merged**